addons/binding/org.openhab.binding.linuxinput/src/main/java/org/openhab/binding/linuxinput/internal/DeviceReadingHandler.java

```diff
@@ -0,0 +1,123 @@
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Abstract handler, that encapsulates the lifecycle of an underlying device.
+ *
+ * @author Thomas Weißschuh - Initial contribution
+ */
+public abstract class DeviceReadingHandler extends BaseThingHandler {
+    private static final Logger LOGGER = LoggerFactory.getLogger(DeviceReadingHandler.class);
+
+    private Future<Void> worker = null;
+
+    abstract boolean immediateSetup() throws IOException;
+
+    abstract boolean delayedSetup() throws IOException;
+
+    abstract void handleEventsInThread() throws IOException;
+
+    abstract void closeDevice() throws IOException;
+
+    public DeviceReadingHandler(Thing thing) {
+        super(thing);
+    }
+
+    @Override
+    public final void initialize() {
+        boolean performDelayedSetup = performImmediateSetup();
+        if (performDelayedSetup) {
+            scheduler.execute(() -> {
+                boolean handleEvents = performDelayedSetup();
+                if (handleEvents) {
+                    worker = scheduler.submit(() -> {
+                        handleEventsInThread();
+                        return null;
+                    });
+                }
+            });
+        }
+    }
+
+    private boolean performImmediateSetup() {
+        try {
+            return immediateSetup();
+        } catch (IOException e) {
+            handleSetupError(e);
+            return false;
+        }
+    }
+
+    private boolean performDelayedSetup() {
+        try {
+            return delayedSetup();
+        } catch (IOException e) {
+            handleSetupError(e);
+            return false;
+        }
+    }
+
+    private void handleSetupError(Exception e) {
+        LOGGER.error("ERROR", e);
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+    }
+
+    @Override
+    public final void dispose() {
+        LOGGER.info("disposing");
+        try {
+            stopWorker();
+        } catch (ExecutionException | TimeoutException e) {
+            LOGGER.error("Interrupted while waiting for worker", e);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        }
+        try {
+            closeDevice();
+        } catch (IOException e) {
+            LOGGER.error("Could not close device", e);
+        }
+        LOGGER.info("disposed");
+    }
+
+    private void stopWorker() throws InterruptedException, ExecutionException, TimeoutException {
+        LOGGER.info("interrupting worker {}", worker);
+        if (worker == null) {
+            return;
+        }
+        worker.cancel(true);
+        try {
+            worker.get(30, TimeUnit.SECONDS);
+        } catch (CancellationException e) {
+            /* expected */
+        }
+        LOGGER.info("worker interrupted");
+        assert worker.isDone();
```

Comment:
Don't use `assert`s in production code there is no guarantee that they will be enable so they could cause non-deterministic behaviour.